src/Workspaces/Core/MSBuild.BuildHost/Rpc/RpcServer.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Reflection;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Rpc
+{
+    /// <summary>
+    /// Implements the server side of the RPC channel used to communicate with the build host.
+    /// </summary>
+    /// <remarks>
+    /// The RPC system implemented here is pretty close to something like JSON-RPC; however since we need the Build Host to be usable in Source Build
+    /// scenarios, we are limited to using only what is either in .NET or can be easily made buildable in Source Build. Thus existing solutions like StreamJsonRpc 
+    /// are out. If at some point there is a standard RPC mechanism exposed in .NET, we should just delete this and use that instead.
+    /// </remarks>
+    internal sealed class RpcServer
+    {
+        private readonly TextWriter _sendingStream;
+        private readonly SemaphoreSlim _sendingStreamSemaphore = new SemaphoreSlim(initialCount: 1);
+        private readonly TextReader _receivingStream;
+
+        private readonly ConcurrentDictionary<int, object> _rpcTargets = new();
+        private volatile int _nextRpcTargetIndex = -1; // We'll start at -1 so the first value becomes zero
+
+        public RpcServer(Stream sendingStream, Stream receivingStream)
+        {
+            _sendingStream = new StreamWriter(sendingStream, Encoding.UTF8);
+            _receivingStream = new StreamReader(receivingStream, Encoding.UTF8);
```

Comment:
it is terrifying to me that an RPC streaming system specifies an encoding in any way.  What happens if we need to send things like null bytes etc?  is something ensuring that everything in this system is pure, legal, text from end to end?

Reply:
This is the text encoding _of the JSON representation._ So any nulls in the text are serialized as a JSON string with embedded escaping.

Reply:
And tests have also been added to cover the null/unicode/newline cases.